README.md
# XOlympics #

Xolympics: a 2 player game where you are the triangle trying to hit the ball to the enemys backboard.

blue controls: up, down, left, and right keys.
red controls: "w", "s", "a", and "d" keys

How to use?
-----------

Xolympics is part of the Sugar desktop.  Please refer to;

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
* [How to use Sugar](https://help.sugarlabs.org/),
* [Download Xolympics](https://activities.sugarlabs.org/en-US/sugar/addon/4259).

Reporting Bugs
--------------

Bugs can be reported in the
[issues tab](https://github.com/sugarlabs/XOlympics/issues)
of this repository.

Contributing
------------

Please consider [contributing](https://github.com/sugarlabs/sugar-docs/blob/master/src/contributing.md) to the project with your ideas and your code.


NEWS
3

* include license, port to GTK+ 3, updated sugargame, fixed bugs

2

* Updated to a newer version of PyBox2D


activity.py
from gettext import gettext as _

from sugar3.activity import activity
from sugar3.graphics.toolbarbox import ToolbarBox
from sugar3.activity.widgets import ActivityToolbarButton
from sugar3.graphics.toolbutton import ToolButton
from sugar3.activity.widgets import StopButton

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk
import tools
import sugargame.canvas
import pygame
import xolympics

class XOlympicsActivity(activity.Activity):

    def __init__(self, handle):
        activity.Activity.__init__(self, handle)

        self.game = xolympics.XOlympicsGame()
        self.game.canvas = sugargame.canvas.PygameCanvas(self, main=self.game.run, modules=[pygame.display, pygame.font])
        self.set_canvas(self.game.canvas)
        self.game.canvas.grab_focus()
        self.build_toolbar()

    # setup the toolbar
    def build_toolbar(self):
        toolbar_box = ToolbarBox()
        self.set_toolbar_box(toolbar_box)
        toolbar_box.show()

        activity_button = ActivityToolbarButton(self)
        toolbar_box.toolbar.insert(activity_button, -1)
        activity_button.show()
       
        separator = Gtk.SeparatorToolItem()
        separator.props.draw = False
        separator.set_expand(True)
        toolbar_box.toolbar.insert(separator, -1)
        separator.show()

        stop_button = StopButton(self)
        toolbar_box.toolbar.insert(stop_button, -1)
        stop_button.show()

        self.show_all()

    def radioClicked(self,button):
        evt = pygame.event.Event(pygame.USEREVENT)
        pygame.event.post(evt)


xolympics.py
#!/usr/bin/python
"""
This file is part of the 'Physics' Project
Physics is a 2D Physics Playground for Kids (supporting Box2D2)
Physics Copyright (C) 2008, Alex Levenson, Brian Jordan
Elements Copyright (C) 2008, The Elements Team, <elements@linuxuser.at>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.



Wiki:   http://wiki.laptop.org/wiki/Physics
IRC:    #olpc-physics on irc.freenode.org

Code:   http://dev.laptop.org/git?p=activities/physics
        git clone git://dev.laptop.org/activities/physics

License:  GPLv3 http://gplv3.fsf.org/
"""

import sys
import math
import pygame
from pygame.locals import *
from pygame.color import *
import elements
from elements import Elements
import tools
from helpers import *
import sys
sys.path.insert(0, "elements")
import box2d
import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk

class XOlympicsGame:
    def __init__(self):
        pass

    def run(self):
        self.rightscore = self.leftscore = 0
        self.forcespeed = 75
        self.jumpforce = 20
        self.leftDPress = False
        self.rightDPress = False
        self.leftLPress = False
        self.leftRPress = False
        self.leftJump = False
        self.rightLPress = False
        self.rightRPress = False
        self.rightJump = False
        self.updateList = []

        pygame.init()
        self.screen = pygame.display.get_surface()
        # get everything set up
        self.clock = pygame.time.Clock()
        self.font = pygame.font.Font(None, 24) # font object
        self.joystickobject = None 
        self.debug = True
        # kids laptop
        # create the name --> instance map for components
        self.toolList = {}
        for c in tools.allTools:
            self.toolList[c.name] = c(self)
        #self.currentTool = self.toolList[tools.allTools[0].name]
       # no tools, eh? 
        # set up the world (instance of Elements)
        self.world = elements.Elements(self.screen.get_size())
        self.world.renderer.set_surface(self.screen)
        # set up static environment
        self.world.set_color((0, 255, 0))  
        self.world.add.ground()
        self.ball = self.world.add.ball((600, 0), 50)
        # ADD LEFT BORDER
        self.world.set_color((255, 0, 0))  
        self.world.add.rect((0,-20), 25, 900, dynamic=False, density=1.0, restitution=0.16, friction=0.5)   
        self.leftplayer = self.world.add.poly(( 264.0, 81.0 ),  ((-109.9405166666667, -64.244016666666653), (110.60718333333335, -63.089316666666605), (-0.66666666666668561, 127.33333333333337)) , dynamic=True, density=1.0, restitution=0.16, friction=0.5, screenCoord=False)
        # ADD RIGHT BORDER
        self.world.set_color((0, 0, 255))  
        self.world.add.rect((1180,-20), 25, 900, dynamic=False, density=1.0, restitution=0.16, friction=0.5)   
        self.rightplayer = self.world.add.poly(( 885.0, 78.0 ),  [(108.94051666666667, -65.976066666666611), (2.6666666666666288, 127.33333333333337), (-111.60718333333341, -61.357266666666646)] , dynamic=True, density=1.0, restitution=0.16, friction=0.5, screenCoord=False)
        # we're getting 2 grounds - grey and green. wtf.  
#	self.leftplayer = 
#        self.world.add.poly((900,800),((-300,300), (300, 300), (300, -300)), dynamic=True, density=1.0, restitution=0.16, friction=0.5)
        # self.leftplayer = self.world.add.rect((500,0), 25, 90, dynamic=True, density=1.0, restitution=0.16, friction=0.5)
        self.leftplayer.linearDamping = 0.07
        self.test = self.leftplayer.worldCenter 
#        self.rightplayer = self.world.add.rect((900,775), 25, 90, dynamic=True, density=1.0, restitution=0.16, friction=0.5)
        # hack fix: set_color early!

        self.running = True    
        while self.running:
            while Gtk.events_pending():
                Gtk.main_iteration()

            for event in pygame.event.get():
                if (event.type == KEYDOWN and (event.key == K_a or event.key == K_KP4)):
                    self.leftLPress = True
                if (event.type == KEYUP and (event.key == K_a or event.key == K_KP4)):
                    self.leftLPress = False
                if (event.type == KEYDOWN and (event.key == K_s or event.key == K_KP2)):
                    self.leftDPress = True
                if (event.type == KEYUP and (event.key == K_s or event.key == K_KP2)):
                    self.leftDPress = False
                if (event.type == KEYDOWN and (event.key == K_d or event.key == K_KP6)):
                    self.leftRPress = True
                if (event.type == KEYUP and (event.key == K_d or event.key == K_KP6)):
                    self.leftRPress = False
                if (event.type == KEYDOWN and (event.key == K_w or event.key == K_KP8)):
                    self.leftJump = True
                if (event.type == KEYUP and (event.key == K_w or event.key == K_KP8)):
                    self.leftJump = False
                if (event.type == KEYDOWN and (event.key == K_LEFT or event.key == K_KP7)):
                    self.rightLPress = True
                if (event.type == KEYUP and (event.key == K_LEFT or event.key == K_KP7)):
                    self.rightLPress = False
                if (event.type == KEYDOWN and (event.key == K_RIGHT or event.key == K_KP1)):
                    self.rightRPress = True
                if (event.type == KEYUP and (event.key == K_RIGHT or event.key == K_KP1)):
                    self.rightRPress = False
                if (event.type == KEYDOWN and (event.key == K_UP or event.key == K_KP9)):
                    self.rightJump = True
                if (event.type == KEYUP and (event.key == K_UP or event.key == K_KP9)):
                    self.rightJump = False            
                if (event.type == KEYDOWN and (event.key == K_DOWN or event.key == K_KP3)):
                    self.rightDPress = True
                if (event.type == KEYUP and (event.key == K_DOWN or event.key == K_KP3)):
                    self.rightDPress = False
                if (event.type == pygame.QUIT):
                    self.running = False
                    pygame.quit()
                    sys.exit()
#            for event in pygame.event.get():
                #self.currentTool.handleEvents(event)
            if self.leftLPress:
                self.leftplayer.ApplyForce(box2d.b2Vec2(-self.forcespeed,0), self.leftplayer.worldCenter, True)
            if self.leftRPress:
                self.leftplayer.ApplyForce(box2d.b2Vec2(self.forcespeed,0), self.leftplayer.worldCenter, True)
            if self.leftJump:
                if self.leftplayer.worldCenter.y < 0.80:
                    self.leftplayer.ApplyLinearImpulse(box2d.b2Vec2(0,self.jumpforce), self.leftplayer.worldCenter, True)
            if self.rightLPress:
                self.rightplayer.ApplyForce(box2d.b2Vec2(-self.forcespeed,0), self.rightplayer.worldCenter, True)
            if self.rightRPress:
                self.rightplayer.ApplyForce(box2d.b2Vec2(self.forcespeed,0), self.rightplayer.worldCenter, True)
            if self.rightDPress:
	            self.rightplayer.ApplyLinearImpulse(box2d.b2Vec2(0,-self.jumpforce), self.rightplayer.worldCenter, True)
            if self.rightJump:
                if self.rightplayer.worldCenter.y < 0.80:
	                self.rightplayer.ApplyLinearImpulse(box2d.b2Vec2(0,self.jumpforce), self.rightplayer.worldCenter, True)
            if self.leftDPress:
	            self.leftplayer.ApplyLinearImpulse(box2d.b2Vec2(0,-self.jumpforce), self.leftplayer.worldCenter, True)
            #if self.leftleft == True
            #    self.leftplayer.ApplyForce((50,0), self.leftplayer.worldCenter)
            # Clear Display
            if self.ball.worldCenter.x < 1:
                print "Goal Blue!", self.rightscore
                self.leftscore += 1
                self.world.set_color((0, 0, 255))
                self.ball = self.world.add.ball((600, 0), 50)
            elif self.ball.worldCenter.x > 11:
                print "Goal Red!", self.rightscore
                self.rightscore += 1
                self.world.set_color((255, 0, 0))
                self.ball = self.world.add.ball((600, 0), 50)

            # For some reason this isn't being reached... thats
            # a problem. THe screen is gray.
            self.screen.fill((255,255,255)) 
            # Update & Draw World
            self.world.update()
            self.world.draw()
            
            # draw output from tools
            #self.currentTool.draw()
            
            #Print all the text on the screen
            
		    #text = self.font.render("Current Tool: "+self.currentTool.name, True, (255,255,255))
            #textpos = text.get_rect(left=700,top=7)
            # self.screen.blit(text,textpos)  
            # for displaying text ^
            # Flip Display
            pygame.display.flip()  
            
            # Try to stay at 30 FPS
            self.clock.tick(30) # originally 50    

def main():
    toolbarheight = 75
    tabheight = 45
    pygame.init()
    pygame.display.init()
    x,y  = pygame.display.list_modes()[0]
    screen = pygame.display.set_mode((x,y-toolbarheight))#-tabheight))
    # create an instance of the game
    game = XOlympicsGame(screen) 
    # start the main loop
    game.run()

# make sure that main get's called
if __name__ == '__main__':
    main()



tools.py
#==================================================================
#                           Physics.activity
#                              Tool Classes
#                           By Alex Levenson
#==================================================================
import pygame
from pygame.locals import *
from helpers import *
from inspect import getmro
from elements import box2d
# tools that can be used superlcass
class Tool(object):
    name = "Tool"
    icon = "icon"
    toolTip = "Tool Tip"
    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Tool"
    def handleEvents(self,event):
        handled = True
        # default event handling
        if event.type == QUIT or (event.type == KEYDOWN and event.key == K_ESCAPE):
            # bye bye! Hope you had fun!
            self.game.running = False
        elif event.type == KEYDOWN:
            if event.key == K_SPACE:
                #space pauses
                self.game.world.run_physics = not self.game.world.run_physics  
            elif event.key == K_t:
                self.game.setTool("triangle")
            elif event.key == K_b:
                self.game.setTool("box")
            elif event.key == K_c:
                self.game.setTool("circle")
            elif event.key == K_j:
                self.game.setTool("joint")
            elif event.key == K_p:
                self.game.setTool("polygon")
            elif event.key == K_g:
                self.game.setTool("grab")
            elif event.key == K_d:
                self.game.setTool("destroy")
            elif event.key == K_m:
                self.game.setTool("magicpen")
            #elif event.key == K_g:
            #    self.game.setTool("gear")
            # Game/joystick-related keys
#            elif event.key == K_KP4: # Left gamepad, left arrow
#                self.game.leftplayer.ApplyForce(box2d.b2Vec2(-500, 0), self.game.leftplayer.GetWorldCenter())
  
   # if self.game.joystickobject:
               #     self.game.joystickobject[0].ApplyTorque(9000) 
               # elif self.game.debug: print "Left gamepad left arrow error: no joystick object selected"
            #elif event.key == K_KP6: # Left gamepad, right arrow
#                self.game.leftplayer.ApplyForce(box2d.b2Vec2(500, 0), self.game.leftplayer.GetWorldCenter())
            elif event.key == K_KP8: # Left gamepad, up arrow
                if self.game.joystickobject:
                    self.game.joystickobject[0].ApplyTorque(9000) 
                elif self.game.debug: print "Left gamepad up arrow error: no joystick object selected"
            elif event.key == K_KP2: # Left gamepad, down arrow
                if self.game.joystickobject:
                    self.game.joystickobject[0].ApplyTorque(9000) 
                elif self.game.debug: print "Left gamepad down arrow error: no joystick object selected"
            #elif event.key == K_KP7: # Right gamepad, square 
#                self.game.rightplayer.ApplyForce(box2d.b2Vec2(-500, 0), self.game.leftplayer.GetWorldCenter())
            #elif event.key == K_KP1: # Right gamepad, check
#                self.game.rightplayer.ApplyForce(box2d.b2Vec2(500, 0), self.game.leftplayer.GetWorldCenter())
            elif event.key == K_KP9: # Right gamepad, circle
                if self.game.joystickobject:
                    self.game.joystickobject[0].ApplyTorque(-9000) 
                elif self.game.debug: print "Right gamepad circle button error: no joystick object selected"
            elif event.key == K_KP3: # Right gamepad, X
                if self.game.joystickobject:
                    self.game.joystickobject[0].ApplyTorque(-9000) 
                elif self.game.debug: print "Right gamepad X button error: no joystick object selected"
        elif event.type == USEREVENT:
            if hasattr(event,"action"):
                if self.game.toolList.has_key(event.action): self.game.setTool(event.action)
        elif event.type == MOUSEBUTTONDOWN and event.button == 1:
            self.game.canvas.grab_focus()
            handled = False
        else:
            handled = False
        return handled                                     
    def draw(self):
        # default drawing method is don't draw anything
        pass
    def cancel(self):
        # default cancel doesn't do anything
        pass

# The circle creation tool        
class CircleTool(Tool): 
    name = "circle"
    icon = "circle"
    toolTip = "Circle"
   
    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Circle"
        self.pt1 = None
        self.radius = None
    def handleEvents(self,event):
        #look for default events, and if none are handled then try the custom events 
        if not super(CircleTool,self).handleEvents(event):
            if event.type == MOUSEBUTTONDOWN:
                if event.button == 1:
                    self.pt1 = pygame.mouse.get_pos()                    
            elif event.type == MOUSEBUTTONUP:
                if event.button == 1:                    
                    if self.radius > 1: # elements doesn't like tiny shapes :(
                        self.game.world.add.ball(self.pt1,self.radius, dynamic=True, density=1.0, restitution=0.16, friction=0.5)                    
                    self.pt1 = None        
                    self.radius = None            
    def draw(self):
        # draw a circle from pt1 to mouse
        if self.pt1 != None:
            self.radius = distance(self.pt1,pygame.mouse.get_pos())
            if self.radius > 3: 
                thick = 3
            else:
                thick = 0
            pygame.draw.circle(self.game.screen, (100,180,255),self.pt1,self.radius,thick) 
            pygame.draw.line(self.game.screen,(100,180,255),self.pt1,pygame.mouse.get_pos(),1)  
    def cancel(self):
        self.pt1 = None  
        self.radius = None
   
# The box creation tool        
class BoxTool(Tool):    
    name = "box"
    icon = "box"
    toolTip = "Box"

    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Box"
        self.pt1 = None        
        self.rect = None
    def handleEvents(self,event):
        #look for default events, and if none are handled then try the custom events 
        if not super(BoxTool,self).handleEvents(event):
            if event.type == MOUSEBUTTONDOWN:
                if event.button == 1:
                    self.pt1 = pygame.mouse.get_pos()                    
            elif event.type == MOUSEBUTTONUP:
                if event.button == 1 and self.pt1!=None:                                 
                    if self.rect.width > 10 and self.rect.height > 10: # elements doesn't like small shapes :(
                        self.game.world.add.rect(self.rect.center, self.rect.width/2, self.rect.height/2, dynamic=True, density=1.0, restitution=0.16, friction=0.5)                   
                    self.pt1 = None        
                                
    def draw(self):
        # draw a box from pt1 to mouse
        if self.pt1 != None:
            width = pygame.mouse.get_pos()[0] - self.pt1[0]
            height = pygame.mouse.get_pos()[1] - self.pt1[1]
            self.rect = pygame.Rect(self.pt1, (width, height))
            self.rect.normalize()           
            pygame.draw.rect(self.game.screen, (100,180,255),self.rect,3)   
    def cancel(self):
        self.pt1 = None  
        self.rect = None      
           
# The triangle creation tool        
class TriangleTool(Tool): 
    name = "triangle"
    icon = "triangle"
    toolTip = "Triangle"
   
    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Triangle"
        self.pt1 = None
        self.vertices = None
    def handleEvents(self,event):
        #look for default events, and if none are handled then try the custom events 
        if not super(TriangleTool,self).handleEvents(event):
            if event.type == MOUSEBUTTONDOWN:
                if event.button == 1:
                    self.pt1 = pygame.mouse.get_pos()                    
            elif event.type == MOUSEBUTTONUP:
                if event.button == 1 and self.pt1!= None:                    
                    if distance(self.pt1,pygame.mouse.get_pos()) > 15: # elements doesn't like tiny shapes :(
                        self.game.world.add.convexPoly(self.vertices, dynamic=True, density=1.0, restitution=0.16, friction=0.5)                    
                    self.pt1 = None
                    self.vertices = None                    
    def draw(self):
        # draw a triangle from pt1 to mouse
        if self.pt1 != None:
            self.vertices = constructTriangleFromLine(self.pt1,pygame.mouse.get_pos())
            pygame.draw.polygon(self.game.screen, (100,180,255),self.vertices, 3) 
            pygame.draw.line(self.game.screen,(100,180,255),self.pt1,pygame.mouse.get_pos(),1)  
    
    def cancel(self):
        self.pt1 = None        
        self.vertices = None

# The Polygon creation tool        
class PolygonTool(Tool):  
    name = "polygon"
    icon = "polygon"
    toolTip = "Polygon"
  
    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Polygon"
        self.vertices = None
    def handleEvents(self,event):
        #look for default events, and if none are handled then try the custom events 
        if not super(PolygonTool,self).handleEvents(event):
            if event.type == MOUSEBUTTONDOWN:
                if event.button == 1:
                    if not self.vertices:
                        self.vertices=[event.pos]  
                    elif distance(event.pos,self.vertices[0]) < 15:                     
                        #self.vertices.append(self.vertices[0]) #connect the polygon
                        gons = decomposePoly(self.vertices)
                        for g in gons:
                            self.game.world.add.convexPoly(g, dynamic=True, density=1.0, restitution=0.16, friction=0.5)
                        #self.game.world.add.complexPoly(self.vertices, dynamic=True, density=1.0, restitution=0.16, friction=0.5)                    
                        self.vertices = None  
                    else:
                        self.vertices.append(event.pos)
                if event.button == 3:
                    if self.vertices:
                        self.vertices.append(event.pos)
                        self.game.world.add.complexPoly(self.vertices, dynamic=True, density=1.0, restitution=0.16, friction=0.5)
                        self.vertices = None
                                
    def draw(self):
        # draw the poly being created
        if self.vertices:
            for i in range(len(self.vertices)-1):
                pygame.draw.line(self.game.screen,(100,180,255),self.vertices[i],self.vertices[i+1],3)
            pygame.draw.line(self.game.screen,(100,180,255),self.vertices[-1],pygame.mouse.get_pos(),3)
            pygame.draw.circle(self.game.screen,(100,180,255),self.vertices[0],15,3)  
    
    def cancel(self):       
        self.vertices = None

# The magic pen tool        
class MagicPenTool(Tool):
    name = "magicpen"
    icon = "magicpen"
    toolTip = "Magic Pen"
    
    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Magic Pen"
        self.vertices = None
    def handleEvents(self,event):
        #look for default events, and if none are handled then try the custom events 
        if not super(MagicPenTool,self).handleEvents(event):
            if event.type == MOUSEBUTTONDOWN:
                if event.button == 1:
                    if not self.vertices:
                        self.vertices=[event.pos]  
                    elif distance(event.pos,self.vertices[0]) < 15:                     
                        self.vertices.append(self.vertices[0]) #connect the polygon
                        self.game.world.add.complexPoly(self.vertices, dynamic=True, density=1.0, restitution=0.16, friction=0.5)                    
                        self.vertices = None  
                    else:
                        self.vertices.append(event.pos)
            elif event.type == MOUSEBUTTONUP:
                if event.button == 1:
                        if self.vertices:
                            self.game.world.add.complexPoly(self.vertices, dynamic=True, density=1.0, restitution=0.16, friction=0.5)                    
                        self.vertices = None
            elif event.type == MOUSEMOTION:
                if self.vertices:
                    self.vertices.append(event.pos)
                                
    def draw(self):
        # draw the poly being created
        if self.vertices:
            for i in range(len(self.vertices)-1):
                pygame.draw.line(self.game.screen,(100,180,255),self.vertices[i],self.vertices[i+1],3)
            pygame.draw.line(self.game.screen,(100,180,255),self.vertices[-1],pygame.mouse.get_pos(),3)
            pygame.draw.circle(self.game.screen,(100,180,255),self.vertices[0],15,3)  
    
    def cancel(self):       
        self.vertices = None

# The grab tool        
class GrabTool(Tool):
    name = "grab"
    icon = "grab"
    toolTip = "Grab"
    
    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Grab"
    def handleEvents(self,event):
        #look for default events, and if none are handled then try the custom events 
        if not super(GrabTool,self).handleEvents(event):
            if event.type == MOUSEBUTTONDOWN:
                if event.button == 1:
                    # grab the first object at the mouse pointer
                    bodylist = self.game.world.get_bodies_at_pos(event.pos, include_static=False) 
                    if bodylist and len(bodylist) > 0:
                        self.game.world.add.mouseJoint(bodylist[0], event.pos)               
            elif event.type == MOUSEBUTTONUP:
                # let it go
                if event.button == 1:
                    self.game.world.add.remove_mouseJoint()
            # use box2D mouse motion
            elif event.type == MOUSEMOTION and event.buttons[0]:
                self.game.world.mouse_move(event.pos)
            
    def cancel(self):
        self.game.world.add.remove_mouseJoint()                        
    
# The joint tool        
class JointTool(Tool):
    name = "joint"
    icon = "joint"
    toolTip = "Joint"
    
    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Joint"
        self.jb1 = self.jb2 = self.jb1pos = self.jb2pos = None
    def handleEvents(self,event):
        #look for default events, and if none are handled then try the custom events 
        if not super(JointTool,self).handleEvents(event):
            if event.type == MOUSEBUTTONDOWN:
                if event.button >= 1:
                    # grab the first body
                    self.jb1pos = event.pos
                    self.jb1 = self.game.world.get_bodies_at_pos(event.pos)
                    self.jb2 = self.jb2pos = None
            elif event.type == MOUSEBUTTONUP:
                if event.button == 1:
                    # grab the second body
                    self.jb2pos = event.pos
                    self.jb2 = self.game.world.get_bodies_at_pos(event.pos)
                    # if we have two distinct bodies, add a distance joint!
                    if self.jb1 and self.jb2 and str(self.jb1) != str(self.jb2):
                        self.game.world.add.distanceJoint(self.jb1[0],self.jb2[0],self.jb1pos,self.jb2pos)                    
                    # If there's only one body, add a fixed joint
                    elif self.jb2:
                        self.game.world.add.fixedJoint(self.jb2[0],self.jb2pos)
                    # regardless, clean everything up
                    self.jb1 = self.jb2 = self.jb1pos = self.jb2pos = None
                if event.button == 3:
                    # add a centered fixed joint
                    '''self.jb2 = self.game.world.get_bodies_at_pos(event.pos)
                    if self.jb2:
                        self.game.world.add.fixedJoint(self.jb2[0])
                    # regardless, clean everything up
                    self.jb1 = self.jb2 = self.jb1pos = self.jb2pos = None'''
                    if self.jb1:
                        self.game.world.add.motor(self.jb1[0],self.jb1pos)
                    self.jb1 = self.jb2 = self.jb1pos = self.jb2pos = None
    def draw(self):
        if self.jb1:
            pygame.draw.line(self.game.screen,(100,180,255),self.jb1pos,pygame.mouse.get_pos(),3)
    
    def cancel(self):
        self.jb1 = self.jb2 = self.jb1pos = self.jb2pos = None             
        
# The destroy tool        
class DestroyTool(Tool):
    name = "destroy"
    icon = "destroy"
    toolTip = "Destroy"
    
    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Destroy"
        self.vertices = None
    def handleEvents(self,event):
        #look for default events, and if none are handled then try the custom events 
        if not super(DestroyTool,self).handleEvents(event):
            if pygame.mouse.get_pressed()[0]:
                if not self.vertices: self.vertices = []
                self.vertices.append(pygame.mouse.get_pos())
                if len(self.vertices) > 10:
                    self.vertices.pop(0)
                tokill = self.game.world.get_bodies_at_pos(pygame.mouse.get_pos())
                if tokill:                        
                        self.game.world.world.DestroyBody(tokill[0])
            elif event.type == MOUSEBUTTONUP and event.button == 1:
                self.cancel()
    def draw(self):
        # draw the trail
        if self.vertices:
            if len(self.vertices) > 1:
                pygame.draw.lines(self.game.screen,(255,0,0),False,self.vertices,3)

    def cancel(self):
        self.vertices = None     

 # The joystick tool        
class JoystickTool(Tool):
    name = "joystick"
    icon = "magicpen" # for now
    toolTip = "Joystick"

    def __init__(self,gameInstance):
        self.game = gameInstance
        self.name = "Joystick"
        self.vertices = None
        #self.joystickobject
    def handleEvents(self,event):
        #look for default events, and if none are handled then try the custom events 
        super(JoystickTool,self).handleEvents(event)
        if pygame.mouse.get_pressed()[0]:
            if not self.vertices:
                self.vertices = []
                self.vertices.append(pygame.mouse.get_pos())
            if len(self.vertices) > 10:
                self.vertices.pop(0)
            self.game.joystickobject = self.game.world.get_bodies_at_pos(pygame.mouse.get_pos())
            if self.game.joystickobject:
                print "hi1"                
        #self.game.world.world.DestroyBody(self.joystickobject[0])
        elif event.type == MOUSEBUTTONUP and event.button == 1:
            self.cancel()
    def draw(self):
        # draw the trail
        if self.vertices:
            if len(self.vertices) > 1:
                pygame.draw.lines(self.game.screen,(255,0,0),False,self.vertices,3)

    def cancel(self):
        self.vertices = None


def getAllTools():
    this_mod = __import__(__name__)
    all = [val for val in this_mod.__dict__.values() if isinstance(val, type)]
    allTools = []
    for a in all:
        if getmro(a).__contains__(Tool) and a!= Tool: allTools.append(a)
    return allTools
            

allTools = getAllTools()
